### ISSUE TITLE
Unable to access flutter web IFrameElement with Canvaskit?

### ISSUE THREAD

**Issue by shah-xad** [flutter/flutter#80195]
I created an Iframe like this **main.dart** :

```dart
import 'package:flutter/material.dart';

import 'dart:async';
import 'dart:html';
import 'dart:io';
import 'dart:js' as js;

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

import 'dart:ui' as ui;

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(

        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  void initState() {
    super.initState();
    _initTeXView();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: HtmlElementView(
          viewType: viewId.toString(),
        ),
      ),
    );
  }

  String viewId = UniqueKey().toString();

  void _initTeXView() {
    ui.platformViewRegistry.registerViewFactory(
        viewId.toString(),
        (int id) => IFrameElement()
          ..width = MediaQuery.of(context).size.width.toString()
          ..height = MediaQuery.of(context).size.height.toString()
          ..src = "https://www.youtube.com/embed/A6XUVjK9W4o"
          ..id = viewId
          ..style.border = 'none');
    js.context.callMethod('accessIFrame', [viewId, "Some Test Data"]);
  }
}
```

And in my **index.html:**

```html
<!DOCTYPE html>
<html>
<head>
  <base href="/">

  <meta charset="UTF-8">
  <meta content="IE=Edge" http-equiv="X-UA-Compatible">
  <meta name="description" content="A new Flutter project.">

  <!-- iOS meta tags & icons -->
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  <meta name="apple-mobile-web-app-title" content="webtestapp">
  <link rel="apple-touch-icon" href="icons/Icon-192.png">

  <!-- Favicon -->
  <link rel="icon" type="image/png" href="favicon.png"/>

  <title>webtestapp</title>
  <link rel="manifest" href="manifest.json">
</head>
<body>
  
  <script>
    if ('serviceWorker' in navigator) {
      window.addEventListener('flutter-first-frame', function () {
        navigator.serviceWorker.register('flutter_service_worker.js');
      });
    }



function accessIFrame(viewId, rawData) {
    var initiated = false;
    document.querySelectorAll("flt-platform-view").forEach(function (platformView) {
            var view = platformView.shadowRoot.children[1];
            if (view.id === viewId) {
                var iframe = view.contentWindow;
                if (iframe != null) {
                  initiated = true;
              
                  console.log("Iframe Found");
                }
            }
        }
    );
    if (!initiated) setTimeout(function () {
      console.log("Try Again");

      accessIFrame(viewId, rawData)
    }, 250);
}
  </script>
  <script src="main.dart.js" type="application/javascript"></script>
</body>
</html>

```

Now This is working fine when I use renderer `html` but doesn't work with `canvaskit`.

**flutter doctor -v**

```console
[✓] Flutter (Channel stable, 2.0.4, on Linux, locale en_US.UTF-8)
    • Flutter version 2.0.4 at /home/shahzad/flutter
    • Framework revision b1395592de (10 days ago), 2021-04-01 14:25:01 -0700
    • Engine revision 2dce47073a
    • Dart version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /home/shahzad/Android/Sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /home/shahzad/Android/Sdk
    • Java binary at: /home/shahzad/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • Ubuntu clang version 11.0.0-2
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.2

[✓] Android Studio
    • Android Studio at /home/shahzad/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.55.1)
    • VS Code at /usr/share/code
    • Flutter extension version 3.21.0

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Linux
    • Chrome (web)    • chrome • web-javascript • Google Chrome 89.0.4389.114

• No issues found!

```

**flutter run -d chrome --web-renderer canvaskit -v**

```console
[  +45 ms] executing: [/home/shahzad/flutter/] git -c log.showSignature=false log -n 1
--pretty=format:%H
[  +27 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[        ] b1395592de68cc8ac4522094ae59956dd21a91db
[        ] executing: [/home/shahzad/flutter/] git tag --points-at
b1395592de68cc8ac4522094ae59956dd21a91db
[  +12 ms] Exit code 0 from: git tag --points-at b1395592de68cc8ac4522094ae59956dd21a91db
[        ] 2.0.4
[  +29 ms] executing: [/home/shahzad/flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[   +7 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/stable
[        ] executing: [/home/shahzad/flutter/] git ls-remote --get-url origin
[   +5 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[  +34 ms] executing: [/home/shahzad/flutter/] git rev-parse --abbrev-ref HEAD
[   +4 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] stable
[  +41 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +1 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[  +45 ms] executing: /home/shahzad/Android/Sdk/platform-tools/adb devices -l
[  +21 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[  +57 ms] executing: /home/shahzad/Android/Sdk/platform-tools/adb devices -l
[   +9 ms] Skipping pub get: version match.
[  +86 ms] Generating
/home/shahzad/Desktop/webtestapp/android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistran
t.java
[  +64 ms] List of devices attached
[   +1 ms] List of devices attached
[  +43 ms] Launching lib/main.dart on Chrome in debug mode...
[  +86 ms] Updating assets
[  +73 ms] Waiting for connection from debug service on Chrome...
[  +62 ms] <- reset
[   +8 ms] /home/shahzad/flutter/bin/cache/dart-sdk/bin/dart --disable-dart-dev
/home/shahzad/flutter/bin/cache/artifacts/engine/linux-x64/frontend_server.dart.snapshot --sdk-root
/home/shahzad/flutter/bin/cache/flutter_web_sdk/ --incremental --target=dartdevc
--debugger-module-names --experimental-emit-debug-metadata -DFLUTTER_WEB_AUTO_DETECT=false
-DFLUTTER_WEB_USE_SKIA=true --output-dill /tmp/flutter_tools.WMQSKD/flutter_tool.KVOVSS/app.dill
--libraries-spec file:///home/shahzad/flutter/bin/cache/flutter_web_sdk/libraries.json --packages
/home/shahzad/Desktop/webtestapp/.dart_tool/package_config.json -Ddart.vm.profile=false
-Ddart.vm.product=false --enable-asserts --track-widget-creation --filesystem-root
/tmp/flutter_tools.WMQSKD/flutter_tools.YKOJTF --filesystem-scheme org-dartlang-app
--initialize-from-dill build/5ce7c659493310451672e1cde9a7cc60.cache.dill.track.dill --platform
file:///home/shahzad/flutter/bin/cache/flutter_web_sdk/kernel/flutter_ddc_sdk.dill
--no-sound-null-safety
[  +11 ms] <- compile org-dartlang-app:/web_entrypoint.dart
[+11454 ms] Waiting for connection from debug service on Chrome... (completed in 11.5s)
[        ] Synced 29.1MB.
[        ] <- accept
[        ] Caching compiled dill
[ +103 ms] Using Google Chrome 89.0.4389.114 

[ +506 ms] [CHROME]:
[        ] [CHROME]:DevTools listening on
ws://127.0.0.1:46377/devtools/browser/4cd01d38-02e4-4068-a08d-3884455b9006
[ +273 ms] DwdsInjector: Received request for entrypoint at
http://localhost:32865/main_module.bootstrap.js
[   +5 ms] MetadataProvider: Loading debug metadata...
[  +13 ms] MetadataProvider: Loaded debug metadata
[   +8 ms] DwdsInjector: Injected debugging metadata for entrypoint at
http://localhost:32865/main_module.bootstrap.js
[+1455 ms] DevHandler: Debug service listening on ws://127.0.0.1:44655/p5jbh2yaNwk=/ws

[   +7 ms] Debug service listening on ws://127.0.0.1:44655/p5jbh2yaNwk=/ws
[        ] Running with unsound null safety
[        ] For more information see https://dart.dev/null-safety/unsound-null-safety
[   +1 ms] 🔥  To hot restart changes while running, press "r" or "R".
[        ] For a more detailed help message, press "h". To quit, press "q".
[+1089 ms] Try Again
[ +241 ms] Try Again
[ +249 ms] Try Again
[ +250 ms] Try Again
[ +251 ms] Try Again
[ +250 ms] Try Again
.
.
.
.
.
.

```

**Comment by iapicca** [comment]
@shah-xad 
can you provide your `flutter doctor -v`, your `flutter run -v` and a complete runnable [minimal reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example)

**Comment by shah-xad** [comment]
I am sorry I'll soon update it.

Althought I found something else. if I render it with  `renderer html` then it works but doesn't work with `renderer canvaskit`, Is there any difference in rendering logic.?

**Comment by iapicca** [comment]
> if I render it with `renderer html` then it works but doesn't work with `renderer canvaskit`

@shah-xad 
this should have been fixed https://github.com/flutter/flutter/issues/69580
make sure you are using the lastest stable or above running
```console
flutter channel stable
flutter upgrade -f
```
if the issue persists please provide your `flutter doctor -v`, your `flutter run -d chrome --web-renderer canvaskit -v`
and a [minimal reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example)

**Comment by shah-xad** [comment]
@iapicca  I have updated my issue please have a look..

**Comment by darshankawar** [comment]
Verified this on latest master and stable and noticed below using the sample code provided:

1. With canvaskit:

```
dhs@Dhss-MacBook-Pro int_test_demo % flutter run -d chrome
Launching lib/main.dart on Chrome in debug mode...
Waiting for connection from debug service on Chrome...             14.6s
Debug service listening on ws://127.0.0.1:52594/Wecpy5fB4xU=/ws

💪 Running with sound null safety 💪

🔥  To hot restart changes while running, press "r" or "R".
For a more detailed help message, press "h". To quit, press "q".
Try Again
Try Again
Try Again
Try Again
```

2. With html:

```
dhs@Dhss-MacBook-Pro int_test_demo % flutter run -d chrome --web-renderer html
Launching lib/main.dart on Chrome in debug mode...
Waiting for connection from debug service on Chrome...             15.1s
Debug service listening on ws://127.0.0.1:52648/Kgk4nZ6U_t0=/ws

💪 Running with sound null safety 💪

🔥  To hot restart changes while running, press "r" or "R".
For a more detailed help message, press "h". To quit, press "q".
Try Again
Iframe Found
```

Looking at the similar issue linked, https://github.com/flutter/flutter/issues/69580 was closed in favor of https://github.com/flutter/flutter/issues/69124 which in turn was fixed and merged a while ago.

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.0.4, on Mac OS X 10.15.4 19E2269 darwin-x64,
    locale en-GB)
    • Flutter version 2.0.4 at /Users/dhs/documents/fluttersdk/flutter
    • Framework revision b1395592de (7 days ago), 2021-04-01 14:25:01 -0700
    • Engine revision 2dce47073a
    • Dart version 2.12.2

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome
[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)    

[✓] VS Code (version 1.55.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0

[✓] Connected device (3 available)
    • SM A260G (mobile)            • 5200763ebcfa861f • android-arm    • Android
      8.1.0 (API 27)
    • Darshan's iphone (mobile) • 21150b119064aecc249dfcfe05e259197461ce23 • ios
      • iOS 14.4.1
    • macOS (desktop)           • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)              • chrome                                   •
      web-javascript • Google Chrome 89.0.4389.114

! Doctor found issues in 1 category.

```

</details>

**Comment by darshankawar** [comment]
Cc: @yjbanov

**Comment by yjbanov** [comment]
It's best to assign a unique `id` in your HTML instead of relying on an internal HTML structure that Flutter creates. We don't provide guarantees about our HTML structure. It is worth noting that this area is changing soon. @ditman  is working on a `<slot>`-based platform view implementation, which should make this use-case more seamless and consistent across renderers. However, the HTML structure will change yet again, so heads up!

**Comment by ditman** [comment]
Please, track progress of the slot implementation here: https://github.com/flutter/flutter/issues/80524

**Comment by ditman** [comment]
I've just confirmed that the new approach will work better here:

https://dit-iframe-reload-example.web.app

Load that app, and see how one can easily do: `document.querySelector('iframe');` without having to worry about shadow roots or anything.

(Source code for the app in #71888, fix described in #80524)
